File: backend/src/main/kotlin/com.vimofthevine.underbudget/model/Account.kt
package com.vimofthevine.underbudget.model

import java.time.Instant
import java.util.UUID

import javax.persistence.*
import javax.validation.constraints.*

import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.beans.factory.annotation.Value
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import org.springframework.data.rest.core.config.Projection
import org.springframework.stereotype.Component

@Entity
@EntityListeners(AuditingEntityListener::class)
@Table(name = "account")
data class Account(
  @Id
  val id: UUID = UUID.randomUUID(),

  @ManyToOne
  @JoinColumn(name = "category_id", nullable = false)
  @get:NotNull(message = "Category is required")
  var category: AccountCategory? = null,

  @get:NotEmpty(message = "Name is required")
  var name: String?,

  @get:NotNull(message = "Institution cannot be null")
  var institution: String? = "",

  @Column(name = "account_number")
  @get:NotNull(message = "Account number cannot be null")
  var accountNumber: String? = "",

  @get:NotNull(message = "Archived cannot be null")
  var archived: Boolean? = false,

  @Column(name = "external_id")
  @get:NotNull(message = "External ID cannot be null")
  var externalId: String? = "",

  @Column(updatable = false)
  @CreatedDate
  var created: Instant?,

  @Column(name = "last_updated")
  @LastModifiedDate
  var lastUpdated: Instant?
)

@Projection(name = "accountOverview", types = arrayOf(Account::class))
interface AccountOverview {
  fun getId(): UUID
  fun getName(): String
  fun isArchived(): Boolean
}

File: backend/underbudget/schemas/account.py
""" Account schema """
from marshmallow import Schema, fields, validate


class AccountSchema(Schema):
    """ Account schema """

    id = fields.Integer(dump_only=True)
    name = fields.String(required=True, validate=validate.Length(min=1))
    institution = fields.String(missing="")
    account_number = fields.String(data_key="accountNumber", missing="")
    archived = fields.Boolean(missing=False)
    external_id = fields.String(data_key="externalId", missing="")
    created = fields.DateTime(dump_only=True)
    last_updated = fields.DateTime(data_key="lastUpdated", dump_only=True)


class AccountCategorySchema(Schema):
    """ Account category schema """

    id = fields.Integer(dump_only=True)
    name = fields.String(required=True, validate=validate.Length(min=1))
    accounts = fields.List(
        fields.Nested(AccountSchema, only=["id", "name", "archived"]),
        dump_only=True
    )
    created = fields.DateTime(dump_only=True)
    last_updated = fields.DateTime(data_key="lastUpdated", dump_only=True)
